Load information from `.arena` file on map load to display it on map loading screen

illwieckz opened the pull request:
Load information from `.arena` file on map load to display it on map loading screen.

Requires this to display the data:

- https://github.com/UnvanquishedAssets/unvanquished_src.dpkdir/pull/161

Fixes: 

- https://github.com/Unvanquished/Unvanquished/issues/2114

illwieckz commented:
I also created a `cg.mapLongName` variable to be used both by the map loading screen and the score board.

This variable uses the long name from the `.arena` file, or fallback on the worldswpawn's message key from the `.bsp` file (what was previously used in the scoreboard), or fallback on the `.bsp` base name.

This means this PR also fixes map naming on scoreboard on various use case.

illwieckz commented:
Example with:

```
{
	map "plat23"
	longname "Platform 23"
	author "EmperorJack"
	type "tremulous"
}
```

I get:

[![map load screen](https://dl.illwieckz.net/b/unvanquished/wip/map-loading-info/20220926-235341-000.unvanquished-map-load-screen.png)](https://dl.illwieckz.net/b/unvanquished/wip/map-loading-info/20220926-235341-000.unvanquished-map-load-screen.png)

This means people can avoid baking text in level shot, that also means we can translate the sentence (something we can't do in a baked text).

illwieckz commented:
In the future I would like to be able to set multiple `author` key for maps having multiple authors, and the game building `name by author, author, and author` string from it, but for now we can start with a simple `author, author, author` string in `author` key.

illwieckz commented:
On a side note, the original code iterating multiple “arenas” is kept on purpose as some legacy games may have more than one map described in one file, and game would have to read all of the files and all the arena blocks to find a map, which I assume the current code is already doing.

What I want to be make sure to only do is to only look for arena files in currently loaded paks. With non-legacy maps it would only read one file and one block.

DolceTriade commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -31,6 +31,174 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+/*
+===============
+CG_LoadArenasFromFile
+TODO: consider deleting, unless we want to display 'longname' somewhere
```
Remove todo?

illwieckz replied:
This is deleted in another commit of the PR.

DolceTriade commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -31,6 +31,174 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+TODO: consider deleting, unless we want to display 'longname' somewhere
+===============
+*/
+void CG_LoadArenasFromFile( char *filename )
```
Do we want to enforce only one arena per file?

Currently it looks like we can have many arenas in a file. I don't think this is what we want or what we do in practice.

DolceTriade commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -31,6 +31,174 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+CG_LoadArenas
+===============
+*/
+void CG_LoadArenas()
```
Let's just do 
```
void CG_LoadArena() {
  CG_LoadArenaFromFile(Str::Format("meta/%s.arena", map));
}
```

DolceTriade reviewed (commented):
Before I review further, what's the goal for the change?

Do we want to bring back global arena parsing? Are we **only** loading the arena for the current map?

If we are only loading the arena for the current map, I'd like to refactor the code to not parse everything and only parse the current arena and remove all code that is there to support storing the full map list based on the arenas.

illwieckz commented:
@DolceTriade 

Now the code tries `Str::Format("meta/%s/%s.arena", map, map)` and if it fails, fallback on “looking for .arena file” for legacy assets. So this should be fast for ported maps, while keeping it working for legacy assets.

The ability to have more than one map per `.arena` file (which is something that was common in quake 3 scene) is kept, but when we directly load the known meta file, the list is expected to only contain one entry and then be very fast to iterate.
 
I assume it makes things fast while keeping compatibility with legacy assets.

I want to keep compatibility with legacy assets, in fact I was thinking about implementing “multiple files, multiple map per file” old way of doing things as a fallback until I seen this was already implemented. 👍️

It looks like all your comments now had an answer.

DolceTriade commented:
Why is it important to maintain compatibility with unported maps? As far as I can tell, no one uses unported maps, all the maps are ported first and then put on a server.

illwieckz commented:
It's easier to test legacy assets (whatever the game) when there is a reference implementation that can load them in Dæmon.

slipher commented on src/cgame/cg_main.cpp:
```diff
@@ -1253,9 +1318,15 @@ void CG_Init( int serverMessageNum, int clientNum, const glconfig_t& gl, const G
 	// load the new map
 	trap_CM_LoadMap(cgs.mapname);
 
+	/* The map must already be loaded at this point. */
```
By the "map being loaded" what you mean is that configstrings must be received from the server and parsed?

illwieckz replied:
The code fallbacks on a string from the bsp itself.

slipher replied:
And the way you're getting this string is by having the *server* read it from the BSP and the client read it from the network.

illwieckz replied:
That's possible. Do we have a better way to do this?

illwieckz replied:
I just noticed something… if what you say is right, this string was read from the network on every frame when displaying the scoreboard, as this code was previously run when rendering `<levelname>` tag.

The code was doing this:

```sh
static void CG_Rocket_DrawLevelName()
{
	Rocket_SetInnerRML( CG_ConfigString( CS_MESSAGE ), RP_QUAKE );
}
```

Now it does (reusing a global variable set at map load):

```sh
static void CG_Rocket_DrawLevelName()
{

	Rocket_SetInnerRML( cg.mapLongName.c_str(), RP_QUAKE );

}
```

I wonder if such thing could be a cause of:

- https://github.com/Unvanquished/Unvanquished/issues/2043

It would have not be enough or the only one cause as it is known that issue can happen when connecting (so, when no scoreboard is displayed).

slipher replied:
No it is not doing something on the network every frame. The server sends config strings at the beginning and when one of them changes and the client stores the latest values.

slipher commented on src/cgame/cg_local.h:
```diff
@@ -1392,6 +1402,10 @@ struct rocketDataSource_t
 
+	arenaInfo_t arenaList[ MAX_ARENA_MAPS ];
+	int arenaCount;
+	int arenaIndex;
```
arenaIndex is unused

illwieckz replied:
removed.

slipher commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -31,12 +31,195 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+		int numdirs = trap_FS_GetFileListRecursive( "", ".arena", dirlist, sizeof( dirlist ) );
+		char *dirptr = dirlist;
+
+		for ( int i = 0, dirlen = 0; i < numdirs; i++, dirptr += dirlen + 1 )
```
`i` is useless

illwieckz replied:
Isn't it needed to test for `i < numdirs` to stop iterating? (This is what the legacy code does, I just moved `dirlen` declaration within the `for` block).

slipher replied:
Yes, my mistake

slipher commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -31,12 +31,195 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+		}
+	}
+
+	Log::Warn( S_SKIPNOTIFY "%i arenas parsed", cg_numArenas );
```
Don't use `Warn` for something normal. Also skipnotify doesn't work

illwieckz replied:
Right, that was legacy code I just reverted, I made it a Debug.

slipher commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -31,12 +31,195 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+		}
+	}
+
+	qsort( rocketInfo.data.arenaList, rocketInfo.data.arenaCount, sizeof( mapInfo_t ), CG_MapLoadNameCompare );
```
The sorting is useless since this list is only accessed by linear search

illwieckz replied:
Well, that was what the original code did, I can simplify it but all I did there was a revert.

illwieckz replied:
Remove the sorting, we don't use that list anywhere else anyway.

slipher commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -31,12 +31,195 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+	if ( text.size() >= MAX_ARENAS_TEXT )
+	{
+		Log::Warn( "file too large: %s is %i, max allowed is %i",
```
This is a pointless error unless you are trying to strictly enforce arena files to have backwards compatibility with some old software.

illwieckz replied:
Right, I only want to read legacy files targeting legacy engine, people are expected to write non legacy files. I removed the test and the error.

This allowed me to turn the function as a boolean one.

slipher commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -31,12 +31,195 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+	if ( CG_LoadArenasFromFile( arena_file.c_str() ) == 0 )
+	{
+		Log::Warn( "Arena file not found: %s", arena_file );
```
You have duplicated file not found warnings here and inside CG_LoadArenasFromFile

illwieckz replied:
Removed.

illwieckz commented:
I did some rewrite and simplification, turning global things into local things (local to their files) when there was not other files using them, ported some C arrays to C++ `vector` and some `char*` strings to `std::string`.

One thing I haven't rewritten is the `CG_ParseInfos` function and related variables. This looks dark magic to me so I don't want to delay this PR on this bikeshedding. This code is meant to work as is (it's a straight reintroduction of a code we had for a decade) and this is good enough™. I even have the intuition we may even remove the `cg_arenaInfos` array entirely and feed the `arenaList` vector directly, but I don't want to take the risk to break this and delay the merge of the PR because of this. Rewriting this would deserve its own PR.

slipher commented on src/cgame/cg_local.h:
```diff
@@ -1318,6 +1320,7 @@ struct rocketMenu_t
 #define MAX_OUTPUTS 16
 #define MAX_MODS 64
 #define MAX_DEMOS 256
+#define MAX_ARENA_MAPS 128
```
unused?

slipher commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -31,12 +31,180 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+	{
+		arenaInfo_t arenaInfo;
+
+		arenaInfo.name = BG_strdup( Info_ValueForKey( cg_arenaInfos[ i ], "map" ) );
```
Remove strdups

illwieckz commented:
So, after sleeping a bit, the `CG_ParseInfos` function looked more clear to me. I rewrote it a bit and managed to entirely remove the writing to a “Info” key-value strorage format. Not only that means the code now sets the `arenaList` directly instead of filling an “Info” string and then read it to set the `arenaInfo` and fill the `arenaList`, it means the content of `.arena` files are not suffering from restrictions of the “Info” format: we are expected to now be able to use the `\` character in values, meaning we may implement escaping `"` the `\"` way or things like that to also enable this character in values, and we are expected to now be able to allow the `.arena` files to use multiple keys with the same name (similar to the ogg metadata format, see https://github.com/Unvanquished/Unvanquished/issues/2114#issuecomment-1229478131 ).

Note: the proposed `author` key name is known to be used for map authos by Smokin' Guns and Q3 Rally, so it is good to do the same.

slipher commented on src/cgame/cg_main.cpp:
```diff
@@ -1164,6 +1164,87 @@ static void GenerateNavmeshes()
 
+static void SetMapInfo( arenaInfo_t arenaInfo )
+{
+	if ( arenaInfo.longName.empty() )
```
It seems wrong to skip making the authors string because there is no long name

illwieckz replied:
It was a leftover from when I was producing a single “map by author” string, removed.

slipher commented on src/cgame/cg_main.cpp:
```diff
@@ -1253,9 +1334,15 @@ void CG_Init( int serverMessageNum, int clientNum, const glconfig_t& gl, const G
 	// load the new map
 	trap_CM_LoadMap(cgs.mapname);
 
+	/* The map must already be loaded at this point. */
```
This comment still needs to be fixed

illwieckz replied:
This was rewritten and this comment is gone.

DolceTriade reviewed (commented):
lgtm % any more of @slipher's comments

slipher commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -31,12 +31,234 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+		{
+			dirlen = strlen( dirptr );
+
+			Log::Warn( "Loading legacy arena file: %s", dirptr );
```
probably shouldn't be a warning

illwieckz replied:
This line was even a duplicate of a Debug print elsewhere. Now removed.

slipher commented on src/cgame/cg_main.cpp:
```diff
@@ -1205,6 +1272,11 @@ void CG_Init( int serverMessageNum, int clientNum, const glconfig_t& gl, const G
 	cgs.gameState = gameState;
 
+	/* Set level name and authors string to be used by the UI.
+	For some reason it must be done after cgs.gameState is set
```
It's not a mystery - `cgs.gameState` is the array of config strings.

illwieckz replied:
Edited the comment.

slipher commented on src/cgame/cg_main.cpp:
```diff
@@ -1164,6 +1164,73 @@ static void GenerateNavmeshes()
+
+	if ( arenaInfo.longName.empty() )
+	{
+		/* Use worldswpawn's message key from .bsp file
```
typo

illwieckz commented:
So, I squashed all the wip commits about implementing the feature (and the countless rewrites).

I added two commits to “_use `std::string` more_”, what do you think about them?

illwieckz commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -41,21 +41,20 @@ static std::vector<arenaInfo_t> arenaList;
 	while ( true )
 	{
-		token = COM_Parse( &buf );
+		std::string token = COM_Parse( &buf );
```
Well, it makes possible to use easier to read syntax like `str == '"something"` or `str.empty()` instead of a bunch of `strcmp(str, "something") == 0` and others. This is just making code looking better.

DolceTriade replied:
Yes. I think std::string_view is a better way to handle this. std::string makes a copy every time you assign something to it.

illwieckz replied:
It looks like `std::string_view` is C++17, do we have that? On my end it doesn't build.

https://en.cppreference.com/w/cpp/string/basic_string_view

illwieckz commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -41,21 +41,20 @@ static std::vector<arenaInfo_t> arenaList;
 ===============
 */
-static void CG_ParseInfos( const char *buf )
+static void CG_ParseInfos( std::string text )
```
Right. Also what's the difference between `Str::StringRef text` and `std::string &text` ?

DolceTriade replied:
std::string& text passes a reference to the std::string object.

Str::StringRef is wrapper around the pointer to the underlying string data.

DolceTriade replied:
Str::StringRef is equivalent to std::string_view, but std::string_view didn't exist when we wrote Str::StringRef

slipher replied:
StringRef is NOT equivalent to string_view. StringRef holds a pointer to a null-terminated string ("C string") while string_view can be an arbitrary byte sequence. Also string_view is C++17 so it still doesn't exist now :)

DolceTriade replied:
Ah. Right, StringRef points an entire string but string_view can point to arbitrary parts of it.

DolceTriade replied:
Ah. Right, StringRef points an entire string but string_view can point to arbitrary parts of it.

illwieckz commented on src/cgame/cg_gameinfo.cpp:
```diff
@@ -67,38 +66,38 @@ static void CG_ParseInfos( const char *buf )
-			if ( !token[ 0 ] )
+			if ( token.empty() )
 			{
-				strcpy( token, "<NULL>" );
```
Well, this looks to be magical code.

The `COM_ParseExt` function is [there](https://github.com/DaemonEngine/Daemon/blob/13e4aca4e6fcce7f521a8c3df50e380500fe422a/src/engine/qcommon/q_shared.cpp#L571), which seems to return a pointer to some memory in a global variable named `com_token` defined [there](https://github.com/DaemonEngine/Daemon/blob/13e4aca4e6fcce7f521a8c3df50e380500fe422a/src/engine/qcommon/q_shared.cpp#L418).

So here maybe I'm duplicating memory.

DolceTriade replied:
Yes, Maybe you can use std::string_view there?

slipher commented:
StringRef in function arguments is good. The other new changes to use std::string are not so great; this means more unneeded copying and allocations. 

The `strcpy(token, "<NULL>")` thing is indeed a bug; COM_Parse ought to return a const pointer. Instead of that you can do `token = "<NULL>"` with a `const char* token`.

illwieckz commented:
@slipher, @DolceTriade what do you think about my latest commit?

Note: it looks like `string_view` isn't available anyway.

DolceTriade commented:
lgtm % @slipher's comments

necessarily-equal commented:
gentle bump. I can give a review after this is rebased

illwieckz commented:
@slipher I modified the last commit to only turn `std::string` into `Str::StringRef` and not `const char*` into `std::string`.

slipher commented:
Last commit LGTM
